streams/src/test/java/org/apache/kafka/streams/kstream/internals/graph/StreamsGraphTest.java

```diff
@@ -124,6 +125,28 @@ public void shouldNotOptimizeWhenAThroughOperationIsDone() {
 
     }
 
+    @Test
+    public void shouldOptimizeSeveralMergeNodesWithCommonKeyChangingParent() {
+        final StreamsBuilder streamsBuilder = new StreamsBuilder();
+        final KStream<Integer, Integer> parentStream = streamsBuilder.stream("input_topic", Consumed.with(Serdes.Integer(), Serdes.Integer()))
+            .selectKey(Integer::sum);
+
+        final KStream<Integer, Integer> childStream1 = parentStream.mapValues(v -> v + 1);
+        final KStream<Integer, Integer> childStream2 = parentStream.mapValues(v -> v + 2);
+        final KStream<Integer, Integer> childStream3 = parentStream.mapValues(v -> v + 3);
+
+        childStream1
+            .merge(childStream2)
+            .merge(childStream3)
+            .to("output_topic");
```

Comment:
Seems I don't understand the root cause of the bug. Don't we need to add a "key dependent" operation (like aggregation) to trigger a repartition and to merge the nodes to only create one repartition topic?

Reply:
Going back to my comment above, when users specify `OPTIMIZE` the code eagerly makes changes for adjusting the graph for any merge nodes that may require repartitioning.  But in this case, we should do nothing as there is nothing to trigger a repartition.  I'll look to fix this, but I don't see this being a quick fix, so my preference is to do this in a follow-up PR.

Reply:
I think we can remove the `to()` operator to verify if we don't fail with a NPE.